src/new/connection.ts

```diff
@@ -99,6 +113,89 @@ class WalletConnection extends BaseConnection {
       this.websocket.on('subscribe_address', handleSubscribeAddress());
     }
   }
+
+  streamEndHandler() {
+    this.streamController?.abort();
+    this.streamController = null;
+  }
+
+  lockStream(streamId: string): boolean {
+    if (this.streamController === null) {
+      this.streamController = new StreamController(streamId);
+      return true;
+    }
+    return false;
+  }
+
+  startStreamingHistory(id: string, firstIndex: number, xpubkey: string, gapLimit: number = -1) {
+    if (this.streamController?.streamId !== id) {
+      throw new Error('There is an on-going stream, cannot start a second one');
+    }
+    if (this.websocket) {
+      const data = JSON.stringify({
+        id,
+        xpub: xpubkey,
+        type: 'request:history:xpub',
+        'first-index': firstIndex,
+        'gap-limit': gapLimit,
+      });
+      this.websocket.sendMessage(data);
+    }
+  }
+
+  sendManualStreamingHistory(
+    id: string,
+    firstIndex: number,
+    addresses: [number, string][],
+    first: boolean,
+    gapLimit: number = -1
+  ) {
+    if (this.streamController?.streamId !== id) {
+      throw new Error('There is an on-going stream, cannot start a second one');
+    }
+    if (this.websocket) {
+      const data = JSON.stringify({
+        id,
+        first,
+        addresses,
+        type: 'request:history:manual',
+        'first-index': firstIndex,
+        'gap-limit': gapLimit,
+      });
+      this.websocket.sendMessage(data);
+    }
+  }
+
+  async stopStream() {
+    await new Promise<void>((resolve, reject) => {
+      if (this.streamController === null) {
+        // There is no active stream.
+        resolve();
+        return;
+      }
+      // Create a timeout so we do not wait indefinetely
+      // It it reaches here we should reject since something went wrong.
+      const timer = setTimeout(() => {
+        reject();
+      }, 10000);
```

Comment:
Hard coded parameter. I guess you should make it at least a constant in this file.

Reply:
Done

Reply:
Thanks!